kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://tanzanite.cluster.engflow.com/invocations/default/b8827329-ddca-46a9-9b31-a41b550a4bfb?testReportRun=5&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on release-24.1 @ [5d952f80b3e1efe2e9aaed73f1fd68433880fcb7](https://github.com/cockroachdb/cockroach/commits/5d952f80b3e1efe2e9aaed73f1fd68433880fcb7):

Fatal error:

```
panic: test timed out after 4m57s
running tests:
	TestKVNemesisSingleNode_ReproposalChaos (4m47s)

```
Stack: 

```
goroutine 16507 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestKVNemesisSingleNode_ReproposalChaos3368060682
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:328: seed: 2343431574739583493
    kvnemesis_test.go:235: kvnemesis logging to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_30/test.outputs/kvnemesis2888986346
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115598 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [A-kv-transactions C-bug C-test-failure O-robot P-1 T-kv branch-release-23.2]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37246

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14685619?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14685619?buildTab=artifacts#/) on release-24.1 @ [f22887420125d75e23b06d9211023152041df203](https://github.com/cockroachdb/cockroach/commits/f22887420125d75e23b06d9211023152041df203):

Fatal error:

```
panic: test timed out after 4m57s
running tests:
	TestKVNemesisSingleNode_ReproposalChaos (4m45s)

```
Stack: 

```
goroutine 27945 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisSingleNode_ReproposalChaos3848693070
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:328: seed: 6171651256317267464
    kvnemesis_test.go:235: kvnemesis logging to /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis2335936744
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115598 kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed [A-kv-transactions C-bug C-test-failure O-robot P-1 T-kv branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/concurrency/lock_table.go
+++ b/pkg/kv/kvserver/concurrency/lock_table.go
@@ -2297,19 +2297,56 @@ func (kl *keyLocks) isLocked() bool {
 // clearLockHeldBy removes the lock, if held, by the transaction referenced by
 // the supplied ID. It is a no-op if the lock isn't held by the transaction.
 //
+// A boolean indicating whether the lock was held by the transaction or not is
+// also returned.
+//
+// TODO(arul): this method may leave the requests' notion of isPromoting in an
+// incorrect state. As such, it shouldn't be called directly -- replace all
+// usages with releaseLock, and pull this into a closure there instead.
+//
 // REQUIRES: kl.mu to be locked.
-func (kl *keyLocks) clearLockHeldBy(ID uuid.UUID) {
+func (kl *keyLocks) clearLockHeldBy(ID uuid.UUID) bool {
 	e, held := kl.heldBy[ID]
 	if !held {
-		return // nothing to do
+		return false // nothing to do
 	}
 	kl.holders.Remove(e)
 	delete(kl.heldBy, ID)
+	return true
 }
 
 func (kl *keyLocks) clearAllLockHolders() {
 	kl.holders.Init()
 	kl.heldBy = nil
+}
+
+// releaseLock releases the lock, if held, by the supplied transaction.
+//
+// REQUIRES: kl.mu to be locked.
+func (kl *keyLocks) releaseLock(txn *enginepb.TxnMeta) {
+	cleared := kl.clearLockHeldBy(txn.ID)
+	if cleared {
+		// There may be requests in the wait queue that belong to the supplied
+		// transaction. They were considered promoters while the lock was held, but
+		// that's no longer the case. As such, their queueOrder.isPromoting flag is
+		// incorrect, and so is their spot in the lock's wait queue. We can fix this
+		// in two different ways:
+		// 1. Recompute queueOrder.isPromoting for waiting requests that belong to
+		// the transaction that just released its lock and then reorder the wait
+		// queue (if needed).
+		// 2. Release any locking requests that belong to the transaction that just
+		// released its lock. They'll re-scan, re-determine whether they're
+		// promoting or not, and be inserted in the correct spot.
+		//
+		// We choose option 2.
+		//
+		// TODO(arul): Option 1, where we push this complexity into
+		// recomputeWaitQueues, is better. We should switch to that, and in doing
+		// so, get rid of all calls to releaseLockingRequestsFromTxn and replace
+		// them with recomputeWaitQueues. Notably, this includes the call to that
+		// method in the lock acquisition path as well.
+		kl.releaseLockingRequestsFromTxn(txn)
+	}
 }
 
 // lockAcquiredOrDiscovered is called when the supplied lock is successfully
@@ -3156,33 +3193,12 @@ func (kl *keyLocks) tryUpdateLockLocked(
 		return false, false
 	}
 	if up.Status.IsFinalized() {
-		kl.clearLockHeldBy(up.Txn.ID)
+		kl.releaseLock(&up.Txn)
 		if !kl.isLocked() {
 			// The lock transitioned from held to unheld as a result of this lock
 			// update.
 			gc = kl.releaseWaitersOnKeyUnlocked()
 		} else {
-			// In some rare cases (such as replays), there may be requests belonging
-			// to the transaction that just released its lock in the lock's wait
-			// queue. These must be requests that are promoting their lock to
-			// something stronger, as otherwise they wouldn't be in the wait queue
-			// to begin with. However, now that the lock has been released, the
-			// queueOrder.isPromoting determination changes as well. We could handle
-			// this two different ways:
-			// 1. Recompute queueOrder.isPromoting for waiting requests that belong to
-			// the transaction that just released the lock and reorder the wait queue
-			// (if needed).
-			// 2. Release any locking requests that belong to the transaction that
-			// just released its lock. They'll re-scan, re-determine whether they're
-			// promoting or not, and get inserted in the correct spot. We choose the
-			// latter option.
-			//
-			// TODO(arul): The former option, where we push this complexity into
-			// recomputeWaitQueues, is better. We should switch to that, and in doing
-			// so, get rid of all calls to releaseLockingRequestsFromTxn and replace
-			// them with recomputeWaitQueues. Notably, this includes the call to that
-			// method in the lock acquisition path as well.
-			kl.releaseLockingRequestsFromTxn(&up.Txn)
 			// If we're in this branch, it must be the case that there were multiple
 			// shared locks held on this key.
 			//
@@ -3285,7 +3301,7 @@ func (kl *keyLocks) tryUpdateLockLocked(
 	}
 
 	if !isLocked {
-		kl.clearLockHeldBy(txn.ID)
+		kl.releaseLock(txn)
 		if !kl.isLocked() {
 			gc = kl.releaseWaitersOnKeyUnlocked()
 		} else {
@@ -3293,7 +3309,6 @@ func (kl *keyLocks) tryUpdateLockLocked(
 			// above for an explanation. Releasing a lock due to ignored sequence
 			// numbers is no different than releasing a lock because the holder's
 			// transaction has been finalized.
-			kl.releaseLockingRequestsFromTxn(&up.Txn)
 			if kl.holders.Len() == 1 {
 				kl.recomputeWaitQueues(st)
 			} else {
@@ -3604,6 +3619,7 @@ func (kl *keyLocks) tryFreeLockOnReplicatedAcquire(
 		return false /* freed */, false /* mustGC */
 	}
 
+	// TODO(arul): this should call releaseLock as well.
 	kl.clearLockHeldBy(acq.Txn.ID)
 	if kl.isLocked() {
 		return true /* freed */, false /* mustGC */

--- a/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_promotion
+++ b/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_promotion
@@ -1184,3 +1184,235 @@ num=1
    queued locking requests:
     active: false req: 47 promoting: true, strength: Exclusive, txn: 00000000-0000-0000-0000-000000000002
     active: true req: 46, strength: Exclusive, txn: 00000000-0000-0000-0000-000000000001
+
+# ------------------------------------------------------------------------------
+# Regression test for https://github.com/cockroachdb/cockroach/issues/121428.
+# Tests a scenario where there's a mismatch between the lock holder information
+# and requests' notion of their promotion status. This can lead to a deadlock
+# between requests.
+# ------------------------------------------------------------------------------
+
+clear
+----
+num=0
+
+new-request r=req48 txn=txn1 ts=10 spans=exclusive@b
+----
+
+scan r=req48
+----
+start-waiting: false
+
+acquire r=req48 k=b durability=u strength=exclusive
+----
+num=1
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+
+new-request r=req49 txn=txn1 ts=10 spans=intent@a+intent@b
+----
+
+scan r=req49
+----
+start-waiting: false
+
+new-request r=req50 txn=txn2 ts=10 spans=exclusive@a
+----
+
+# For this scenario to be realistic, req49 above must have waited at a different
+# key and dropped its latches. We don't show that as it's not critical to the
+# setup -- we'll simply re-scan with req49 further down the line.
+scan r=req50
+----
+start-waiting: false
+
+acquire r=req50 k=a durability=u strength=exclusive
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+
+new-request r=req51 txn=txn2 ts=10 spans=intent@a+intent@b
+----
+
+scan r=req51
+----
+start-waiting: true
+
+print
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+scan r=req49
+----
+start-waiting: true
+
+print
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+# Both transactions get aborted and release their locks.
+release txn=txn2 span=a
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+release txn=txn1 span=b
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+   queued locking requests:
+    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+# Deadlock.
+
+# ------------------------------------------------------------------------------
+# Very similar to the test above, however, instead of releasing the locks by
+# aborting the transaction, we'll instead roll back sequence numbers.
+# ------------------------------------------------------------------------------
+
+clear
+----
+num=0
+
+new-request r=req52 txn=txn1 ts=10 spans=exclusive@b
+----
+
+scan r=req52
+----
+start-waiting: false
+
+acquire r=req52 k=b durability=u strength=exclusive
+----
+num=1
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+
+new-request r=req53 txn=txn1 ts=10 spans=intent@a+intent@b
+----
+
+scan r=req53
+----
+start-waiting: false
+
+new-request r=req54 txn=txn2 ts=10 spans=exclusive@a
+----
+
+# For this scenario to be realistic, req53 above must have waited at a different
+# key and dropped its latches. We don't show that as it's not critical to the
+# setup -- we'll simply re-scan with req53 further down the line.
+scan r=req54
+----
+start-waiting: false
+
+acquire r=req54 k=a durability=u strength=exclusive
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+
+new-request r=req55 txn=txn2 ts=10 spans=intent@a+intent@b
+----
+
+scan r=req55
+----
+start-waiting: true
+
+print
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+scan r=req53
+----
+start-waiting: true
+
+print
+----
+num=2
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+# Both transactions release their locks by rolling back their sequence numbers.
+update txn=txn2 ts=10 epoch=1 span=a ignored-seqs=0
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+  holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
+   queued locking requests:
+    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+update txn=txn1 ts=10 epoch=1 span=b ignored-seqs=0
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: true req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+ lock: "b"
+   queued locking requests:
+    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+
+# Deadlock.

--- a/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_promotion
+++ b/pkg/kv/kvserver/concurrency/testdata/lock_table/lock_promotion
@@ -1280,8 +1280,7 @@ release txn=txn2 span=a
 num=2
  lock: "a"
    queued locking requests:
-    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
-    active: true req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
  lock: "b"
   holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
    queued locking requests:
@@ -1293,14 +1292,34 @@ release txn=txn1 span=b
 num=2
  lock: "a"
    queued locking requests:
-    active: false req: 51 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
-    active: true req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
  lock: "b"
    queued locking requests:
-    active: false req: 49 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
-    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: false req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
 
-# Deadlock.
+# No Deadlock -- req49 and req51 can scan again without any issue.
+
+scan r=req49
+----
+start-waiting: false
+
+scan r=req51
+----
+start-waiting: true
+
+# Once they do, we see that the wait queues are ordered by sequence number, and
+# that req49 can proceed to evaluation.
+print
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+ lock: "b"
+   queued locking requests:
+    active: false req: 49, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 51, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
 
 # ------------------------------------------------------------------------------
 # Very similar to the test above, however, instead of releasing the locks by
@@ -1395,8 +1414,7 @@ update txn=txn2 ts=10 epoch=1 span=a ignored-seqs=0
 num=2
  lock: "a"
    queued locking requests:
-    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
-    active: true req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
  lock: "b"
   holder: txn: 00000000-0000-0000-0000-000000000001 epoch: 1, iso: Serializable, ts: 10.000000000,0, info: unrepl [(str: Exclusive seq: 0)]
    queued locking requests:
@@ -1408,11 +1426,31 @@ update txn=txn1 ts=10 epoch=1 span=b ignored-seqs=0
 num=2
  lock: "a"
    queued locking requests:
-    active: false req: 55 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
-    active: true req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
  lock: "b"
    queued locking requests:
-    active: false req: 53 promoting: true, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
-    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+    active: false req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
 
-# Deadlock.
+# No Deadlock -- req53 and req55 can scan again without any issue.
+
+scan r=req53
+----
+start-waiting: false
+
+scan r=req55
+----
+start-waiting: true
+
+# Once they do, we see that the wait queues are ordered by sequence number, and
+# that req53 can proceed to evaluation.
+print
+----
+num=2
+ lock: "a"
+   queued locking requests:
+    active: false req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: true req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002
+ lock: "b"
+   queued locking requests:
+    active: false req: 53, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 55, strength: Intent, txn: 00000000-0000-0000-0000-000000000002

